### ISSUE TITLE
Review possible changes

### ISSUE THREAD

**Issue by FlexYourBrain** [defold/template-platformer#2]
## Referencing issue: https://github.com/defold/defold/issues/7473

 Reviewing what changes are needed to update this template to the latest features. ( Open discussion )

### ∙ Issues

**(1)** Users often find that dragging/losing-focus of the game window causes large jumps in delta time making the character move through collision objects.

**(2)** Dragging/Loosing-focus of the game window while using input causes the character to keep moving even when no longer being pressed and window focus is regained.

**(3)** Player hflip/facing based on self.velocity in update every tick. Walking left then releasing input flips character to face right. Not much of an issue, may feel off putting for those learning from it.

**(4)** Physics Debug not showing when update frequency is set to 0 - variable frame rate

**(5)** Warning when using physics debug with this template.

> WARNING:RENDER: Out of debug vertex data (10000). Increase graphics.max_debug_vertices to avoid this warning. Setting graphics.max_debug_vertices

### ∙ Possible Changes

 - player.script update function to use fixed_update instead -- fix issue (1)
 - game.project > Physics > (check) Use Fixed Timestep -- fix issue (1)
 - Use window_callback function for events focus lost/focus gained - release/acquire input -- fix issue (2) except for: window drag (no api for this?)
 - create local bool and logic to hflip/face the character in latest input direction only when needed -- fix issue (3)
 - game.project > Display > (un-check) Vsync and set Update Frequency to 60-- fix issues (4)
 - setting game.project > Graphics > Max Debug Vertices to 140000 stops the warning but why so high? -- Fix issue(5) don't think this is really a fix though? Notes below for closer look.

### ∙ Notes

**Issue 2:**

   I don't think there can be api window events for when mouse-click or dragging the game window title/borders occurs.
    If player drags or holds mouse-click on the window title/borders or when player drags and scales the borders these
    are 2 separate events window_scaling and window_drag at the same time. Defold can detect scaling event but not
    click/drag window. I believe windows stops the game thread while the click drag occurs. (not sure how other OS
    handle this). One problem is that if the player was applying movement input when the click/drag event occurs the
    characters velocity/direction/max_speed/gravity etc. will still be applied when the event is done even when the
    input is no longer being pressed and will stay this way until player presses the input once again to cancel the
    movement then again to move as usual. This can seem like a bug.

   Possible workarounds may be to use mouselock, however that would prevent players from accessing the window at all.
    Another workaround might be to run the game on a seperate thread which may possibly still render the game while
    being dragged not sure though if and how that would work, may not be worth the effort and wouldn't fix the input issue.

**Issue 5:**

  Getting a warning `WARNING:RENDER: Out of debug vertex data (10000)` and setting max debug vertices to 140,000 stops
   the warning but why is it so high? there shouldn't be that many vertices in the scene let alone debug vertices. I
   wouldn't think there would be that much vertices for the debug outlines, maybe it also is counting all vertices in
   the scene. So I took a closer look in renderdoc.

![rendoc_vertex](https://user-images.githubusercontent.com/28400489/228602288-311e94b0-a38e-44dc-885c-bd39ee93c066.png)

**∙ Vertex Count:**
   character sprite = 6
   tile quads =    7,710
   collision  =    15,150
   debug draw =    712

   Total vertices = 23,578

   In conclusion is the debug draw Warning bugged? forcing to set max limit to much much higher number than is actually used? Is this occuring with all projects?

Some of these potential changes may not be needed at all for the scope of this template , noted any issues I found. Any feedback is welcome.

 

> Side note - generated tile collision shapes for square tiles create 4 tris per tile quad(12 vertices), is this due to 1 tri per
> normal direction? Or is it due to a convex shape algorithm? Is this standard or can it be optimized from 12 to 6 same as tiles?

Will review/discuss other templates/tutorials before submitting any pr's

**Comment by britzl** [comment]
Wow, thank you for the thorough investigation. Resizing and dragging of windows is a recurring "problem" that is encountered by users. We should if possible update some of our template projects to better deal with this if possible (perhaps ignoring large delta time values?).

We will have to take a look at the physics debug draw and see if there is a problem or not. I can't say for sure.

> Physics Debug not showing when update frequency is set to 0 - variable frame rate

This was surprising!

**Comment by britzl** [comment]
> **(1)** Users often find that dragging/losing-focus of the game window causes large jumps in delta time making the character move through collision objects.

Changed to a fixed timestep: https://github.com/defold/template-platformer/commit/f77dda687ecd1c6ef483ec1df49b2e1d98142b8f

**Comment by britzl** [comment]
> **(2)** Dragging/Loosing-focus of the game window while using input causes the character to keep moving even when no longer being pressed and window focus is regained.

This is a general issue in the engine. We could listen to window focus changes and set velocity.x to 0, but that would not solve it entirely.

**Comment by britzl** [comment]
> **(3)** Player hflip/facing based on self.velocity in update every tick. Walking left then releasing input flips character to face right. Not much of an issue, may feel off putting for those learning from it.

Added direction of facing: https://github.com/defold/template-platformer/commit/1411137fe274978a8830473954c9e22a9996fedc

**Comment by britzl** [comment]
> **(5)** Warning when using physics debug with this template.

Increased number of debug vertices: https://github.com/defold/template-platformer/commit/f77dda687ecd1c6ef483ec1df49b2e1d98142b8f

**Comment by britzl** [comment]
> **(4)** Physics Debug not showing when update frequency is set to 0 - variable frame rate

Unchecked vsync and set update frequency to 60: https://github.com/defold/template-platformer/commit/1a025938c81fec823f28bba2d776b5dc499fa525

**Comment by britzl** [comment]
@FlexYourBrain Again, many thanks for the great investigation! I've fixed most of the issues, but can I ask that you create a separate ticket in the Defold repo for the physics debug rendering? Thanks.

**Comment by FlexYourBrain** [comment]
I had made made changes and forgotten to submit the PR, sorry about that! Looks good. I will create a ticket for the physics debug issue possibly this weekend.
